ramips: add support for JDCloud RE-CP-02

FH0 opened the pull request:
## Specifications

- Soc: MediaTek MT7621AT
- RAM: 512 MB (DDR3)
- Flash: 16 MB (SPI NOR)
- WiFi: MediaTek MT7905DAN, MediaTek MT7975DN
- Ethernet: 1 WAN, 3 LAN (Gigabit)
- Buttons: Reset, Joylink
- LEDs: (red, blue, green), routed to one indicator in the top of the device
- Power: DC 12V 1A tip positive
- 1 TF Card Slot

The pins for the serial console are already labeled on the board J4 (V, R, T, G). Serial settings: 3.3V, 115200

MAC addresses:

|         | MAC               | Algorithm |
| ------- | ----------------- | --------- |
| label   | dc:d8:xx:xx:xx:01 | label     |
| LAN     | dc:d8:xx:xx:xx:01 | label     |
| WAN     | dc:d8:xx:xx:xx:02 | label+1   |
| WLAN 2g | dc:d8:xx:xx:xx:03 | label+2   |
| WLAN 5g | de:d8:xx:xx:xx:04 | label+3   |

## Installation

1. rename the openwrt-ramips-mt7621-jdcloud_re-cp-02-squashfs-sysupgrade.bin to JDCOS.bin

2. start a TFTP server from IP address 192.168.68.10 and serve the image named JDCOS.bin

3. connect your device to the LAN port

4. power up the router and press any key on the console to interrupt the boot process.

5. enter the following commands on the router console

   1. setenv bootcount 6

   2. saveenv

   3. reset

      > NOTE: wait for the restart, it will automatically fetch the image named JDCOS.bin from the TFTP server and write it into the flash. After the writing is completed, the router will be automatically restarted.

## Known Issues:

Unable to recognize large-capacity TF card, see #14042. But the patch [#issuecomment-1910769942](https://github.com/openwrt/openwrt/issues/14042#issuecomment-1910769942) works

FH0 commented:
make #14487 continue

musashino205 commented:
there are some formal issues:

- please wrap the commit message within 74 characters
- please use the same "`your name <your email>`" for `Signed-off-by: ` and [the author field of the commit (`From: `)](https://github.com/openwrt/openwrt/pull/15372.patch)

ref: https://openwrt.org/submitting-patches

> - commit description
>   - must have less than 75 characters per line
>
> - the `Author` field must match the `Signed-off-by:` line

musashino205 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1524,6 +1524,17 @@ define Device/jcg_y2
+  DEVICE_VENDOR := JD-Cloud
+  DEVICE_MODEL := RE-CP-02
+  DEVICE_PACKAGES := kmod-mt7915-firmware kmod-sdhci-mt7620 \
+                     kmod-fs-ext4
```
If this `kmod-fs-ext4` is unnecessary for basic use of that device, please drop it.

FH0 commented:
`JDCloud RE-CP-02` has a 64GB TF card preinstalled in factory. I want to mount it when boot. It need `kmod-fs-ext4` and `block-mount` packages. And modify `/etc/config/fstab` by following change. I don't know how to make it working.

```diff
3c3
<         option  anon_mount      '0'
---
>         option  anon_mount      '1'
```

```bash
root@OpenWrt:~# df -h
Filesystem                Size      Used Available Use% Mounted on
/dev/root                 4.5M      4.5M         0 100% /rom
tmpfs                   216.4M    200.0K    216.2M   0% /tmp
tmpfs                   216.4M     60.0K    216.3M   0% /tmp/root
tmpfs                   512.0K         0    512.0K   0% /dev
/dev/mtdblock6            8.0M    360.0K      7.6M   4% /overlay
overlayfs:/overlay        8.0M    360.0K      7.6M   4% /
/dev/mmcblk0p4           55.8G    635.4M     54.6G   1% /mnt/mmcblk0p4
```

FH0 commented:
- commit message wrapped
- author matched

DragonBluep commented on target/linux/ramips/dts/mt7621_jdcloud_re-cp-02.dts:
```diff
@@ -0,0 +1,185 @@
+
+&state_default {
+	gpio {
+		groups = "i2c";
```
groups = "uart3", “jtag”, "wdt";

FH0 replied:
Thanks.

FH0 commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -29,6 +29,7 @@ alfa-network,ax1800rm|\
 allnet,all0256n-4m|\
 allnet,all0256n-8m|\
 allnet,all5002|\
+jdcloud,re-cp-02|\
```
Thanks.

DragonBluep commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -1524,6 +1524,16 @@ define Device/jcg_y2
 
+define Device/jdcloud_re-cp-02
+  $(Device/dsa-migration)
+  $(Device/uimage-lzma-loader)
```
By the way. lzma-loader may not be necessary for the WiFi6 device with the new bootloader. If not sure, don't touch it.

FH0 replied:
Yes, I'm not familiar with `uimage-lzma-loader`.

DragonBluep reviewed (commented):
LGTM. If you like, you can add your `Co-authored-by:` tag to the commit message.

shenghuang147 committed:
ramips: add support for JDCloud RE-CP-02

- Soc: MediaTek MT7621AT
- RAM: 512 MB (DDR3)
- Flash: 16 MB (SPI NOR)
- WiFi: MediaTek MT7905DAN, MediaTek MT7975DN
- Ethernet: 1 WAN, 3 LAN (Gigabit)
- Buttons: Reset, Joylink
- LEDs: (red, blue, green), routed to one indicator in the top of the
  device
- Power: DC 12V 1A tip positive
- 1 TF Card Slot

The pins for the serial console are already labeled on the board
J4(V, R, T, G). Serial settings: 3.3V, 115200

MAC addresses:

|         | MAC               | Algorithm |
| ------- | ----------------- | --------- |
| label   | dc:d8:xx:xx:xx:01 | label     |
| LAN     | dc:d8:xx:xx:xx:01 | label     |
| WAN     | dc:d8:xx:xx:xx:02 | label+1   |
| WLAN 2g | dc:d8:xx:xx:xx:03 | label+2   |
| WLAN 5g | de:d8:xx:xx:xx:04 | label+3   |

1. rename the
   openwrt-ramips-mt7621-jdcloud_re-cp-02-squashfs-sysupgrade.bin
   to JDCOS.bin

2. start a TFTP server from IP address 192.168.68.10 and serve the
   image named JDCOS.bin

3. connect your device to the LAN port

4. power up the router and press any key on the console to interrupt
   the boot process.

5. enter the following commands on the router console

   1. setenv bootcount 6

   2. saveenv

   3. reset

      > NOTE: wait for the restart, it will automatically fetch the
      > image named JDCOS.bin from the TFTP server and write it into
      > the flash. After the writing is completed, the router will be
      > automatically restarted.

Unable to recognize large-capacity TF card, see #14042. But the patch
https://github.com/openwrt/openwrt/issues/14042#issuecomment-1910769942
works

Co-Authored-By: Jianti Chen <clbcjt@outlook.com>
Signed-off-by: Sheng Huang <shenghuang147@gmail.com>

FH0 commented:
> LGTM. If you like, you can add your `Co-authored-by:` tag to the commit message.

`Co-authored-by:` added.

hauke commented on target/linux/ramips/dts/mt7621_jdcloud_re-cp-02.dts:
```diff
@@ -0,0 +1,186 @@
+
+&gmac1 {
+	status = "okay";
+	label = "wan";
```
Is this label property working? I think labels are not supported on normal ports only on DSA ports.

DragonBluep replied:
There is a hack patch to support this feature.
https://github.com/openwrt/openwrt/blob/main/target/linux/ramips/patches-6.6/700-net-ethernet-mediatek-support-net-labels.patch

FH0 replied:
Yes. I test it on real device.
```bash
BusyBox v1.36.1 (2024-05-06 01:17:35 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r26181+1-d1504a1e76
 -----------------------------------------------------
root@OpenWrt:~# cat /etc/os-release
NAME="OpenWrt"
VERSION="SNAPSHOT"
ID="openwrt"
ID_LIKE="lede openwrt"
PRETTY_NAME="OpenWrt SNAPSHOT"
VERSION_ID="snapshot"
HOME_URL="https://openwrt.org/"
BUG_URL="https://bugs.openwrt.org/"
SUPPORT_URL="https://forum.openwrt.org/"
BUILD_ID="r26181+1-d1504a1e76"
OPENWRT_BOARD="ramips/mt7621"
OPENWRT_ARCH="mipsel_24kc"
OPENWRT_TAINTS="no-all"
OPENWRT_DEVICE_MANUFACTURER="OpenWrt"
OPENWRT_DEVICE_MANUFACTURER_URL="https://openwrt.org/"
OPENWRT_DEVICE_PRODUCT="Generic"
OPENWRT_DEVICE_REVISION="v0"
OPENWRT_RELEASE="OpenWrt SNAPSHOT r26181+1-d1504a1e76"
root@OpenWrt:~# uname -a
Linux OpenWrt 6.6.29 #0 SMP Mon May  6 01:17:35 2024 mips GNU/Linux
root@OpenWrt:~# ifconfig wan
wan       Link encap:Ethernet  HWaddr 
          inet addr:192.168.1.114  Bcast:192.168.1.255  Mask:255.255.254.0
          inet6 addr:  Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:177737 errors:0 dropped:0 overruns:0 frame:0
          TX packets:113662 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:200370487 (191.0 MiB)  TX bytes:15256169 (14.5 MiB)
          Interrupt:19
```
